### PR TITLE
Support sparse_softmax_cross_entropy_with_logits op with large class

### PR DESCRIPTION
**Pull request by lucienwang1009**
- When the number of label category is too large, the current converter will generate a huge eye matrix which cannot be stored in protobuf. This PR uses `GatherNd` to circumvent the problem.
- Put some complex conversion code (`Range`, `SparseSoftmax...`) into `function/` dir.  
- Wrap `Range` and `GatherNd` conversion functions.

### CODE REVIEW

File: tf2onnx/function/range.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    base_name = utils.make_name(name)
+    start = start_node.get_tensor()
+    limit = limit_node.get_tensor()
+    delta = delta_node.get_tensor()
+    val = np.arange(start, limit, delta, dtype=start.dtype)
+    const_range = ctx.make_const(base_name, val)
+    return ctx.make_node("Identity", [const_range.output[0]], dtypes=[dtype], outputs=[output])
+
+def make_range_subgraph(ctx, start, limit, delta, output, name, dtype):
+    """make Range subgraph."""
```
Comment (nbcsm) [inline tf2onnx/function/range.py]: make_range_non_const

File: tf2onnx/function/range.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    loop_inputs = [trip_count_node.output[0], cond_name, start]
+    loop_node = ctx.make_node("Loop", loop_inputs, output_count=2, op_name_scope=base_name, name="loop",
+                              attr={"body": body_graph})
+    nodes.append(loop_node)
+
+    identity_node = ctx.make_node("Identity", [loop_node.output[1]], name=base_name, dtypes=[dtype], outputs=[output])
+    nodes.append(identity_node)
+
+    return nodes
+
```
Comment (nbcsm) [inline tf2onnx/function/range.py]: also add a make_range, which forward the call to make_range_const or make_range_non_const, based on whether start_node,limit_node, delta_node are all const.

so if someone want to make a range, he does not need to know whether it is const or not, just call make_range.

File: tf2onnx/function/range.py (hunk 1 above)
Comment (nbcsm) [reply tf2onnx/function/range.py]: accept start/limit/delta as node instead of str (output name), to keep the interface consistent.

File: tf2onnx/function/range.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+"""
+tf2onnx.tf2onnx - range op conversion
+"""
+import numpy as np
+from onnx import helper
+from onnx.onnx_pb import TensorProto
+from tf2onnx import utils
+
+# pylint: disable=useless-return,broad-except,logging-not-lazy,unused-argument,missing-docstring
+def make_range_const(ctx,
```
Comment (nbcsm) [inline tf2onnx/function/range.py]: remove these pylint rule unless necessary. i am pretty sure there is no broad-except or useless-return

Comment (nbcsm) [reply tf2onnx/function/range.py]: so for the other files.

File: tf2onnx/tfonnx.py [hunk 4]
```diff
@@ -21,8 +21,10 @@
 
 import tf2onnx
 from tf2onnx import utils
-from tf2onnx.function.select import select_op8
 from tf2onnx.function.gathernd import gathernd_op
+from tf2onnx.function.range import range_op7
+from tf2onnx.function.select import select_op8
+from tf2onnx.function.sparse_softmax_cross_entropy_with_logits import sparse_softmax_cross_entropy_with_logits_op
```
Comment (nbcsm) [inline tf2onnx/tfonnx.py]: this will be verbose if we move more complex ops to functions.

how about put xxxx_op that we want to export into ```tf2onnx/function/__init__.py```
and just ```from tf2onnx.function import *``` here

File: tf2onnx/function/gathernd.py [hunk 5]
```diff
@@ -43,19 +43,16 @@ def make_gathernd_inner_loop(ctx, params, index, dtype):
...
+                               name=name,
                                attr={"body": body_graph})
     nodes.append(inner_loop.op)
     return nodes, inner_loop
 
-def gathernd_op(ctx, node, name, args):
-    """GatherNd op."""
-    # T output = GatherNd(T Input, INT32/INT64 indices)
+
+def make_gathernd_subgraph(ctx, params, indices, output, name, t_params):
```
Comment (nbcsm) [inline tf2onnx/function/gathernd.py]: just make_gathernd?

Comment (nbcsm) [inline tf2onnx/function/gathernd.py]: from make_gathernd perspective, there is no name concept, but scope/prefix, right?
change it to scope_name?

Comment (nbcsm) [reply tf2onnx/function/gathernd.py]: so are the others.

File: tf2onnx/function/gathernd.py [hunk 6]
```diff
@@ -86,17 +83,21 @@ def gathernd_op(ctx, node, name, args):
...
+                    make_onnx_inputs_outputs(result_name, t_params, [])]
     body_nodes = []
     index = ctx.make_node("Gather", [flatten_indices.output[0], trip_name], attr={"axis": 0})
     index_squeeze = ctx.make_node("Squeeze", [index.output[0]], attr={"axes": [0]})
     # inner loop to gather result
-    inner_loop_nodes, inner_loop = make_gathernd_inner_loop(ctx, params, index_squeeze, node_dtype)
+    inner_loop_nodes, inner_loop = make_gathernd_inner_loop(ctx,
+                                                            params,
+                                                            index_squeeze,
+                                                            "{}_inner_loop".format(name),
```
Comment (nbcsm) [inline tf2onnx/function/gathernd.py]: make_name to make sure it is unique?

Comment (nbcsm) [reply tf2onnx/function/gathernd.py]: so is name="{}_loop".format(name)

File: tests/test_backend.py [hunk 7]
```diff
@@ -1410,6 +1410,20 @@ def test_sparse_softmax_cross_entropy_with_logits(self):
 
         self._run_test_case([_OUTPUT], {_INPUT: label_val, _INPUT1: logits_val})
 
+    @unittest.skipIf(*support_op_with_target('rs6', 'SparseSoftmaxCrossEntropyWithLogits'))
+    def test_sparse_softmax_cross_entropy_with_logits_large_class(self):
+        num_class = 30000
```
Comment (nbcsm) [inline tests/test_backend.py]: make sure it can pass locally

File: tests/test_backend.py [hunk 8]
```diff
@@ -1410,6 +1410,20 @@ def test_sparse_softmax_cross_entropy_with_logits(self):
 
         self._run_test_case([_OUTPUT], {_INPUT: label_val, _INPUT1: logits_val})
 
+    @unittest.skipIf(*support_op_with_target('rs6', 'SparseSoftmaxCrossEntropyWithLogits'))
```
Comment (nbcsm) [inline tests/test_backend.py]: on a second thought, i feel it is kind of ambiguous to use 'rs6' , for both converter code and test.
what we really want is when backend is onnxruntime and its version equals to xxx, do some special hacks in conversion.

we may introduce "onnxruntime" as a target, and enhance current target handling logic to support version (optional).
e.g. 
--target onnxruntime
--target onnxruntime:0.1.4

and in test, we can skip if backend==onnxruntime and target != onnxruntime.

Comment (nbcsm) [reply tests/test_backend.py]: for this pr, just keep it this way.
we can create new pr if we want to add onnxruntime as target.

Comment (nbcsm) [reply tests/test_backend.py]: @pengwa to comment

File: tests/test_backend.py (hunk 7 above)
Comment (lucienwang1009) [reply tests/test_backend.py]: sure

File: tf2onnx/function/range.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+tf2onnx.tf2onnx - range op conversion
+"""
+import numpy as np
+from onnx import helper
+from onnx.onnx_pb import TensorProto
+from tf2onnx import utils
+
+# pylint: disable=unused-argument,missing-docstring
+def make_range_const(ctx, start_node, limit_node, delta_node, output, name, dtype):
+    """make Range subgraph if all inputs are const."""
```
Comment (nbcsm) [inline tf2onnx/function/range.py]: make make_range_const and make_range_non_const internal.

Comment (nbcsm) [reply tf2onnx/function/range.py]: or make make_range_const and make_range_non_const internal.

File: tf2onnx/function/range.py [hunk 10]
```diff
@@ -89,13 +89,10 @@ def make_range_non_const(ctx, start, limit, delta, output, name, dtype):
...
-def make_range(ctx, start, limit, delta, output, name, dtype):
-    start_node = ctx.get_node_by_output(start)
-    limit_node = ctx.get_node_by_output(limit)
-    delta_node = ctx.get_node_by_output(delta)
-    if start_node.is_const() and limit_node.is_const() and delta_node.is_const():
-        return make_range_const(ctx, start_node, limit_node, delta_node, output, name, dtype)
-    return make_range_non_const(ctx, start, limit, delta, output, name, dtype)
+def make_range(ctx, start, limit, delta, output, scope_name, dtype):
+    if all([ctx.get_node_by_output(n).is_const() for n in [start, limit, delta]]):
+        return make_range_const(ctx, start, limit, delta, output, scope_name, dtype)
```
Comment (nbcsm) [inline tf2onnx/function/range.py]: all([ctx.get_node_by_output(n).is_const() for n in [start, limit, delta]])
=>
all(ctx.get_node_by_output(n).is_const() for n in [start, limit, delta])

no need to create a list again.

File: tf2onnx/function/gathernd.py [hunk 11]
```diff
@@ -5,20 +5,21 @@
...
 
-# pylint: disable=useless-return,broad-except,logging-not-lazy,unused-argument,missing-docstring
+# pylint: disable=unused-argument,missing-docstring
 
-INT64_MAX = np.iinfo(np.int64)
+INT64_MAX = np.iinfo(np.int64).max
 
-def make_gathernd_inner_loop(ctx, params, index, dtype):
+def make_gathernd_inner_loop(ctx, params, index, name, dtype):
     """create the inner loop for GatherNd."""
```
[the remark this answers is not in the document]
Comment (nbcsm) [reply tf2onnx/function/gathernd.py]: i mean make_gathernd_inner_loop...

**Comment by lucienwang1009** [comment]
@pengwa plea have a look though it's not very urgent.

File: tf2onnx/function/gathernd.py [hunk 12]
```diff
@@ -105,6 +107,7 @@ def gathernd_op(ctx, node, name, args):
     gathernd_loop = ctx.make_node("Loop",
                                   [outter_shape_sum.output[0], cond_const.output[0], params],
                                   output_count=2,
+                                  name="{}_loop".format(scope_name),
```
Comment (pengwa) [inline tf2onnx/function/gathernd.py]: you can use op_name_scope=scope_name, name=loop.

File: tf2onnx/function/range.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    # T range = Range(T start, T limit, T delta)
+    # V v_final_and_scan_outputs = Loop(int64 M, B cond, V v_initial)
+    base_name = utils.make_name(scope_name)
+    start = ctx.get_node_by_output(start).get_tensor()
+    limit = ctx.get_node_by_output(limit).get_tensor()
+    delta = ctx.get_node_by_output(delta).get_tensor()
+    val = np.arange(start, limit, delta, dtype=start.dtype)
+    const_range = ctx.make_const(base_name, val)
+    return ctx.make_node("Identity", [const_range.output[0]], dtypes=[dtype], outputs=[output])
+
```
Comment (pengwa) [inline tf2onnx/function/range.py]: two lines between non-class functions.

File: tf2onnx/function/gathernd.py [hunk 14]
```diff
@@ -130,11 +133,22 @@ def gathernd_op(ctx, node, name, args):
...
-                                   outputs=[node.output[0]])
+                                   outputs=[output])
     nodes.extend([inner_loop_shape,
                   inner_loop_shape_,
                   output_inner_shape,
                   output_shape_,
                   output_shape,
                   output_reshape])
     return nodes
+
```
Comment (pengwa) [inline tf2onnx/function/gathernd.py]: one more line

File: tf2onnx/function/sparse_softmax_cross_entropy_with_logits.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+
+"""
+tf2onnx.tf2onnx - sparse_softmax_cross_entropy_with_logits op conversion
+"""
+import numpy as np
+from onnx.onnx_pb import TensorProto
+from tf2onnx import utils
+from tf2onnx.function.range import make_range
+from tf2onnx.function.gathernd import make_gathernd
+
```
Comment (pengwa) [inline tf2onnx/function/sparse_softmax_cross_entropy_with_logits.py]: one more line

File: tf2onnx/tfonnx.py [hunk 16]
```diff
@@ -205,91 +204,6 @@ def identity_op(ctx, node, name, args):
     return node
```
Comment (pengwa) [inline tf2onnx/tfonnx.py]: cool!
